Import Resources by Identifiers (Name, etc.)

denniskniep opened the pull request:
### Description of your changes
There is a feature request around how to import objects via <realm/name> instead of < uuid >

 see: https://github.com/crossplane-contrib/provider-keycloak/issues/126

There are already some workarounds in place:
1. Using terraform `import` property (only a few resources supports that!)
2. Using crossplane functions to import the built-in objects of keycloak (see [here](https://github.com/crossplane-contrib/provider-keycloak?tab=readme-ov-file#functions-an-compositions))

This is a proposal how to import resources by name. 

We could use a custom implementation of `ExternalName` config (see [here](https://github.com/crossplane/upjet/blob/main/docs/images/upjet-externalname.png) and [here](https://github.com/crossplane/upjet/blob/main/docs/configuring-a-resource.md))

* `GetIDFn`: Use defined props to lookup the Id used in the specific keycloak instance for that resource  (i.e. by realmId, clientId & roleId). This could be done by using the terraform projects´s `keycloakClient` (i.e. `kcClient.GetRoleByName(ctx, realmId.(string), clientId.(string), name.(string))`)
* `GetExternalNameFn`: Builds the ExternalName based on deterministic props (i.e. `<realmId>/<clientId>/<roleId>`)

What do you think about this approach? 

**Topics to clarify:**
- [ ]  External Name and some props are overlapping. In case of role resource realmId, clientId & roleId are specified in properties as well as in external name. As far as I know there is a concept to mitigate that: `OmittedFields` & `SetIdentifierArgumentFn`. Issue with that is, that we can´t specify ClientIDRef which is a big downside!
- [ ] Renewing tf-state creates new resource instead of updating existing when renaming one of the identifying properties

Fixes #126

Breee commented:
i like the idea. 

Importing by `<realmId>/<clientId>/<roleId>`  would be perfect,   especially if there are no weird keycloak UUIDS involved.  

Changing this might break this function:

https://gitlab.com/corewire/images/crossplane/function-keycloak-builtin-objects   

but we have to test it

denniskniep commented:
@Breee I stumbled upon an case where the change lead to probably unwanted/unexpected behavior:

When one of the identifying properties are changed i.e. `realmId` or `clientId` in the `OidcClient`.
and
the provider pod is not currently running and started after that change _(or the tf-state in workspace is renewed for any different reason)_
then
a new OidcClient will be created and the old is left over.

Examples:
1. when provider pod is running and therefore tf-state is persisted in workspace. Renaming ClientId property from `abc` to `123` will result in an update of the existing client
2. when provider pod is not running. Renaming ClientId property from `abc` to `123`. Then starting the provider pod will result in an newly created client `123` and the old client `abc` still exists.

What if we still maintain the UUID´s in external-name, but if UUID is empty or not existing, then utilize the identifying properties to retrieve the UUID?

We could apply the following logic:
1. Check if external-name is set
2. If external-name is set: try to resolve resource by UUID (get UUID from external-name)
3. If resource can be resolved by UUID: return UUID
4. If resource can NOT be resolved by UUID OR external-name is NOT set: try to resolve resource by identifying properties (realmId, clientId, etc.)
5.  If resource can be resolved by identifying properties: return UUID
6.  return indicator to create resource, because it does not exists

Breee commented:
yes

> @Breee I stumbled upon an case where the change lead to probably unwanted/unexpected behavior:
> 
> When one of the identifying properties are changed i.e. `realmId` or `clientId` in the `OidcClient`. and the provider pod is not currently running and started after that change _(or the tf-state in workspace is renewed for any different reason)_ then a new OidcClient will be created and the old is left over.
> 
> Examples:
> 
> 1. when provider pod is running and therefore tf-state is persisted in workspace. Renaming ClientId property from `abc` to `123` will result in an update of the existing client
> 2. when provider pod is not running. Renaming ClientId property from `abc` to `123`. Then starting the provider pod will result in an newly created client `123` and the old client `abc` still exists.
> 
> What if we still maintain the UUID´s in external-name, but if UUID is empty or resource with that UUID is not existing, then utilize the identifying properties to retrieve the UUID?
> 
> We could apply the following logic:
> 
> 1. Check if external-name is set
> 2. If external-name is set: try to resolve resource by UUID (get UUID from external-name)
> 3. If resource can be resolved by UUID: return UUID
> 4. If resource can NOT be resolved by UUID OR external-name is NOT set: try to resolve resource by identifying properties (realmId, clientId, etc.)
> 5. If resource can be resolved by identifying properties: return UUID
> 6. return indicator to create resource, because it does not exists

Yes, good catch. If people mess with the objects in keycloak that's probably in more cases a problem (it's a shame that they can not be made immutable on the UI) 

I like the logic you proposed.  

We then also have to test, how this affects existing deployments

denniskniep commented:
@Breee  seems to be that `make generate` throws errors when I am using generated types in `config` module:

```
	r := x.RoleParameters{}
	err = utils.UnmarshalTerraformParamsToObject(parameters, &r)
	if err != nil {
		return "", err
	}

	if r.RealmID == nil {
		return "", errors.New("realmId not set")
	}

	if r.Name == nil {
		return "", errors.New("name not set")
	}

	clientId := ""
	if r.ClientID != nil {
		clientId = *r.ClientID
	}
```

Any ideas how to solve that? Else I will switch back to my previous approach:

```
        realmID, realmIdExists := parameters["realm_id"]
	if !realmIdExists {
		return "", errors.New("realmId not set")
	}

	name, nameExists := parameters["name"]
	if !nameExists {
		return "", errors.New("name not set")
	}

	clientID, clientIdExists := parameters["client_id"]
	if !clientIdExists {
		clientID = ""
	}
```

...then we won´t get any compiler errors once those used attribute names are changed. 
1. But I guess that is a very rare case for identifying properties.
2. Thats already the case in `provider.go.KnownReferencers()`

denniskniep commented:
@Breee 
I am getting closer...Only the ldap-resources are missing.

Can you already have a look at it?
BTW: It was necessary to access a private KeycloakClient method, because not every exposed API is implemented by the client. I would like to tackle this once we have a final version. Maybe terraform-keycloak-provider is willing to accept those additional methods.

How should we proceed? Does it make sense to release a new major version along with that feature to highlight the change?
Furthermore we could release it as ReleaseCandidate/PreRelease i.e "2.0.0-rc.1" and improve it further, but already ask for input from the community and offer users an opportunity to experiment with that changed approach.

What do you think?

Breee commented:
Yeah i think a RC would be a good idea.  
Do you think this branch is ready or do you need more testing?

denniskniep commented:
> Yeah i think a RC would be a good idea. Do you think this branch is ready or do you need more testing?

@Breee okay cool, I already added handling for the missing ldap resources and added support for end to end tests.
Before we release a first RC I want to add further end to end tests

consooo commented on dev/setup_dev_environment.sh:
```diff
@@ -190,26 +192,26 @@ export KEYCLOAK_PASSWORD=admin
+$kubectl_cmd apply -f ${SCRIPT_DIR}/apps/keycloak-provider/keycloak-provider-config.yaml
 fi
 
 if [[ "$uselocalprovider" == "true" ]]; then
```
in l. 218, can you change the line to `$kubectl_cmd apply -f "${SCRIPT_DIR}/../package/crds"` otherwise --use-local-provider won't work when the script is run from root dir.

denniskniep commented:
Created e2e testcases are running via uptest

Existing import test case:
* state and status.id is removed, external-name is left

New import test cases:
* state and status.id is removed, external-name is removed
* state and status.id is removed, external-name is set to "Incorrect"

denniskniep commented on config/authentication/config.go:
```diff
@@ -79,3 +83,139 @@ func Configure(p *config.Provider) {
+// ExecutionIdentifierFromIdentifyingProperties is used to find the existing resource by it´s identifying properties
+var ExecutionIdentifierFromIdentifyingProperties = lookup.BuildIdentifyingPropertiesLookup(executionIdentifyingPropertiesLookup)
+
+func getExecutionIDByExternalName(ctx context.Context, id string, parameters map[string]any, kcClient *keycloak.KeycloakClient) (string, error) {
```
This follows the signature from crossplane provided ID functions, I would prefer to stick to that

denniskniep commented:
@Breee now we are currently internally testing this in our enviroment. I think we can release it on Monday or Tuesday as RC

denniskniep commented:
@Breee we completed our tests, I think this branch is ready for a RC

Breee commented:
@denniskniep  we invited to be a maintainer of the repo.  Can you recreate your branch in this repo? then we can create a RC

Breee commented:
` xpkg.upbound.io/crossplane-contrib/provider-keycloak:v2.0.0-rc.1`  is the pre-release  

@b509  do you probably also want to test this out?

Breee commented:
i guess this is ready to go, what do you think @denniskniep ?

denniskniep commented:
@Breee  yes, tests are completed in our environment.
I guess we can move forward.

What we should note is, that if someone is using `import: true`. This could be set to `false` now. If someone wants to apply that change, then the affected resources demand recreation. Else no changes are necessary at resource level.

Breee commented:
Let' s merge and create a release.  
you can then write Excessive information into the release to highlight the new feature fully
